lib/grandpa/votes_tracker.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"container/list"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// votesTracker tracks vote messages that could
+// not be processed, and removes the oldest ones once
+// its maximum capacity is reached.
+// It is NOT THREAD SAFE to use.
+type votesTracker struct {
+	// map of vote block hash to authority ID (ed25519 public Key)
+	// to data (peer id + message + tracking linked list element pointer)
+	mapping map[common.Hash]authorityIDToData
+	// double linked list of block hash + authority ID
+	// to track the order vote messages were added in.
+	linkedList *list.List
+	capacity   int
+}
+
+type authorityIDToData map[ed25519.PublicKeyBytes]voteMessageMapData
+
+type voteMessageMapData struct {
+	peerID  peer.ID
+	message *VoteMessage
+	// element contains a blockHashAuthID value which
+	// itself contains a block hash and an authority ID.
+	element *list.Element
+}
+
+type blockHashAuthID struct {
+	blockHash   common.Hash
+	authorityID ed25519.PublicKeyBytes
+}
+
+// newVotesTracker creates a new vote message tracker
+// with the capacity specified.
+func newVotesTracker(capacity int) votesTracker {
+	return votesTracker{
+		mapping:    make(map[common.Hash]authorityIDToData, capacity),
+		linkedList: list.New(),
+		capacity:   capacity,
+	}
+}
+
+func newBlockHashAuthID(blockHash common.Hash,
+	authorityID ed25519.PublicKeyBytes) blockHashAuthID {
+	return blockHashAuthID{
+		blockHash:   blockHash,
+		authorityID: authorityID,
+	}
+}
+
+// add adds a vote message to the vote message tracker.
+// If the vote message tracker capacity is reached,
+// the oldest vote message is removed.
+func (vt *votesTracker) add(peerID peer.ID, voteMessage *VoteMessage) {
+	signedMessage := voteMessage.Message
+	blockHash := signedMessage.BlockHash
+	authorityID := signedMessage.AuthorityID
+
+	voteMessages, has := vt.mapping[blockHash]
+	if !has {
+		// add new block hash in tracker
+		vt.cleanup()
+		elementData := newBlockHashAuthID(blockHash, authorityID)
+		element := vt.linkedList.PushFront(elementData)
+		data := voteMessageMapData{
+			peerID:  peerID,
+			message: voteMessage,
+			element: element,
+		}
+		vt.mapping[blockHash] = authorityIDToData{
+			authorityID: data,
+		}
+		return
```

Comment:
Isn't this code exactly the same as lines 100 to 108?

Reply:
Yes, I simplified it to avoid repeating it. Good catch :+1: